find_lr.py

```diff
@@ -1,56 +1,74 @@
 from baseline.datasets import CasiaSurfDataset
-from torchvision import transforms
-from torch import optim, nn
 from torch.utils import data
 import torch
 import matplotlib.pyplot as plt
 from argparse import ArgumentParser
 import models
-import numpy as np
+import losses
+import optimizers
+import math
 from tqdm import tqdm
-from baseline.datasets import NonZeroCrop
+import yaml
 
+from transforms import TrainTransform
 
-def get_loss(lr, dataloader, args):
-    model = getattr(models, args.model)(num_classes=args.num_classes)
-    model = model.to(device)
-    model.train()
-    optimizer = optim.Adam(model.parameters(), lr=lr)
-    loss_fn = nn.CrossEntropyLoss()
+
+def find_lr(loader, net, optimizer, criterion, init_value=3e-10, final_value=3., beta=0.98):
+    num = len(loader) - 1
+    mult = (final_value / init_value) ** (1 / num)
+    lr = init_value
+    optimizer.param_groups[0]['lr'] = lr
+    avg_loss = 0.
+    best_loss = 0.
     losses = []
-    for i, (images, labels) in enumerate(dataloader):
-        if i > 4:
-            return np.mean(losses)
-        images, labels = images.to(device), labels.to(device)
-        outputs = model(images)
-        loss = loss_fn(outputs, labels)
-        losses.append(loss.item())
+    log_lrs = []
+    for batch_num, data in enumerate(tqdm(loader)):
+        # As before, get the loss for this mini-batch of inputs/outputs
+        inputs, labels = data
+        optimizer.zero_grad()
+        outputs = net(inputs)
+        loss = criterion(outputs, labels)
+        # Compute the smoothed loss
+        avg_loss = beta * avg_loss + (1 - beta) * loss.item()
+        smoothed_loss = avg_loss / (1 - beta ** (batch_num + 1))
+        # Stop if the loss is exploding
+        if batch_num > 0 and smoothed_loss > 4 * best_loss:
+            return log_lrs, losses
+        # Record the best loss
+        if smoothed_loss < best_loss or batch_num == 0:
+            best_loss = smoothed_loss
+        # Store the values
+        losses.append(smoothed_loss)
+        log_lrs.append(math.log10(lr))
+        # Do the SGD step
         loss.backward()
         optimizer.step()
+        # Update the lr for the next step
+        lr *= mult
+        optimizer.param_groups[0]['lr'] = lr
+    return log_lrs, losses
 
 
 if __name__ == '__main__':
     argparser = ArgumentParser()
-    argparser.add_argument('--model', required=True, type=str)
-    argparser.add_argument('--num_classes', type=int, default=2)
-    argparser.add_argument('--depth', type=bool, default=False)
-    argparser.add_argument('--ir', type=bool, default=False)
+    argparser.add_argument('--config_path', required=True, type=str)
+    argparser.add_argument('--optimizer', default="Adam", type=str)
```

Comment:
Looks like we have a duplication here - optimizer is already written in config file.